src/devices/sound/hc55516.h

```diff
@@ -1,95 +1,210 @@
 // license:BSD-3-Clause
-// copyright-holders:Aaron Giles
+// copyright-holders:Aaron Giles, Jonathan Gevaryahu
+// thanks-to:Zonn Moore
 #ifndef MAME_SOUND_HC55516_H
 #define MAME_SOUND_HC55516_H
 
 #pragma once
 
-class hc55516_device : public device_t, public device_sound_interface
+class cvsd_device : public device_t, public device_sound_interface
 {
 public:
-	hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
-
-	/* sets the digit (0 or 1) */
+	enum
+	{
+		RISING=true,
+		FALLING=false
+	};
+	cvsd_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// common things across all the CVSD chips:
+	// Clock pull, really only relevant of something manually polls the clock (and clock is specified) for some reason, which is a very bad design pattern
+	// this function WILL ASSERT if it is called and the clock hz is NOT specified!
+	READ_LINE_MEMBER( clock_r );
+	// A clock state change callback, because where we're going we don't need sanity
+	auto clock_state_cb() { return m_clock_state_push_cb.bind(); }
+
+	// Clock push
+	// this function WILL ASSERT if it is called and the clock hz IS specified!
+	WRITE_LINE_MEMBER( mclock_w );
+
+	// Digital in pull callback function, for use if a clock is specified and we need to pull in the digital pin state, otherwise unused
+	auto digin_cb() { return m_digin_pull_cb.bind(); }
+	// Digital in push to the pin, as a pseudo 'buffer'
+	WRITE_LINE_MEMBER( digin_w );
+
+	// Audio In pin, an analog value (int16_t -32768 - 32767, or a float?) of the audio waveform being pushed to the chip
+	void audio_in_w(int16_t data); // TODO: handle encoding
+
+	// DEC/ENC decode/encode select push
+	WRITE_LINE_MEMBER( dec_encq_w ); //TODO: handle encoding
+
+	// Digital out push callback function
+	auto digout_cb() { return m_digout_push_cb.bind(); }
```

Comment:
Keep configuration member functions grouped together and separate from live functions.

Reply:
I made some changes here, but I'm not sure if this is exactly what you wanted or not.

Reply:
I believe this should be resolved.